backend/gsr_booking/api_wrapper.py

```diff
@@ -532,38 +311,175 @@ def get_affiliation(self, email):
+                booking = GSRBooking.objects.create(
+                    user_id=curr_user.id,
+                    booking_id=str(booking_id),
+                    gsr=gsr,
+                    room_id=rid,
+                    room_name=room_name,
+                    start=curr_start,
+                    end=curr_end
+                )            
+                booking.reservation = reservation
+                booking.save()
+
+                if (curr_start := curr_end) >= end:
+                    break
+        except APIError as e:
+            raise APIError(f"{str(e)}. Was only able to book {start.strftime('%H:%M')} - {curr_start.strftime('%H:%M')}")
+
+        return reservation
+    
+    def cancel_room(self, booking_id, user):
+        # I think this should prefetch reservation, reservation__gsrbooking_set, and gsr ?
+        if gsr_booking := GSRBooking.objects.filter(booking_id=booking_id).prefetch_related(Prefetch('reservation__gsrbooking_set'), Prefetch('gsr')).first():
+            if gsr_booking.user != user and gsr_booking.reservation.creator != user:
                 raise APIError("Error: Unauthorized: This reservation was booked by someone else.")
+            
+            (self.WBW.cancel_room if gsr_booking.gsr.kind == GSR.KIND_WHARTON else self.LBW.cancel_room)(booking_id, gsr_booking.user)
+            
             gsr_booking.is_cancelled = True
             gsr_booking.save()
-        response = self.request("POST", f"{API_URL}/1.1/space/cancel/{booking_id}").json()
-        if "error" in response[0]:
-            raise APIError("LibCal: " + response[0]["error"])
-        return response
 
-    def check_credits(self, user, lc_start=None):
-        # default to beginning of day
-        if lc_start is None:
-            lc_start = make_aware(datetime.datetime.now())
-            lc_start = lc_start.replace(second=0, microsecond=0, minute=0, hour=0)
-
-        lc_end = lc_start + datetime.timedelta(days=1)
-
-        # filters for all reservations for the given date
-        reservations = GSRBooking.objects.filter(
-            gsr__in=GSR.objects.filter(kind=GSR.KIND_LIBCAL),
-            start__gte=lc_start,
-            end__lte=lc_end,
-            is_cancelled=False,
-            user=user,
-        )
-        total_minutes = 0
-        for reservation in reservations:
-            # accumulates total minutes over all reservations
-            total_minutes += int((reservation.end.timestamp() - reservation.start.timestamp()) / 60)
-        # 120 minutes per day
-        return {CreditType.LIBCAL.value: 120 - total_minutes}
+            reservation = gsr_booking.reservation
+            if all(booking.is_cancelled for booking in reservation.gsrbooking_set.all()):
+                reservation.is_cancelled = True
+                reservation.save()                
+        else:
+            try: 
```

Comment:
Perhaps something like this? Nested try-excepts kinda look weird imo lol:
```
services = [self.WBW, self.LBW]

for service in services:
    try:
        service.cancel_room(booking_id, user)
        break
    except APIError:
        continue

raise APIError("Error: Unknown booking id")
```

Reply:
smart smart